### ISSUE TITLE
Define remote relation methods immediately at setup time

### ISSUE THREAD

**Issue by bajtos** [strongloop/loopback#2448]
At the moment, LoopBack defers definition of remoting metadata for relation methods until the model is connected to a datasource. This is problematic in cases where we want to gather remoting metadata without connecting to databases, typically in `apic loopback:refresh`.

The proposal is to modify `PersistedModel` to always define remoting metdata for relation methods immediately at `setup` time, the same way as we are already defining remoting metadata for DAO methods at `setup` time.

**Comment by jannyHou** [comment]
To make sure I understand this issue correctly, let me explain more about our current design:


In our current implementation, the REST build-in methods for persisted models `setRemoting` immediately in [`PersistedModel.setup`(lb)](https://github.com/strongloop/loopback/blob/master/lib/persisted-model.js#L44) phase, which completes in [`defineModels`(lb-boot)](https://github.com/strongloop/loopback-boot/blob/master/lib/executor.js#L196)

But at this phase, relations/scopes are **NOT** loaded yet, they are defined when attaching model to app  [`app.model`(lb-boot)](https://github.com/strongloop/loopback-boot/blob/master/lib/executor.js#L202) and [setupDataAccess](https://github.com/strongloop/loopback-datasource-juggler/blob/master/lib/datasource.js#L541) for it. 
So we create a relation/scope resolver function in [Model.setup(lb)](https://github.com/strongloop/loopback/blob/1a62ed7f277db82e0b70371540af6a7bf7146b8c/lib/model.js#L236) but defer its execution until we setup component Explorer, which is after relations/scopes loaded.

The work flow looks like: 
```js
// First setupModels in lb-boot

defineModel() {
  setupModel() {
    // Persisted models' DAO methods are loaded now
    // setRemoting for them
  }
}
attachModelToApp() {
   attachModelToDatasource() {
      setupDataAccess() {
       // DefineRelations
       // Here we get the function of each relation method, and wrap them into a function for setRemoting
      }
  }
}

//  Then setupComponents in lb-boot

setupExplorer() {
  ExplorerRequiresSwaggerObject() {
    SharedClassLoadResolver() {
      // Here we setRemoting for relation/scope method 
    }
  }
}

```


> LoopBack defers definition of remoting metadata for relation methods until the model is connected to a datasource

Since we already have lazyConnect when booting from swaggerGenerator, I assume the `connected` here  means `attachTo`.

But IMO we are not able to define them before attaching to ds because some code in relationDefinition defines properties directly to datasource, [see example](https://github.com/strongloop/loopback-datasource-juggler/blob/master/lib/relation-definition.js#L611)

To summarize, the tricky thing is we cannot separate the following two functions:
```js
relationDefinition() {
  modelName.datasource.defineProperty();
  warpRelationMethodForRemoting();
}
```

**Comment by bajtos** [comment]
@jannyHou the trick is that you don't need relations to be resolved (relation methods to be defined) in order to declare remoting metadata. Take a look at what we do in `lib/persisted-model.js` - we declare remoting metadata for dummy functions throwing an error. Only when the model is attached to a datasource, these dummy method are replaced with a full implementation. The same approach should work for relation methods too. Also notice that in order to expose relation methods via strong-remoting, we have wrapper methods defined on the model itself, e.g. `product.categories.create` has a method `product.__create__category` that's invoked remotely.

**Comment by jannyHou** [comment]
I had a chat with Miroslav regarding this issue. Here is a brief conclusion:
To `setRemoting` for relations defined in `model.json` and dynamically defined in code, we need to grab metadata from `Model.relations[ ]`, which has to be set after model attached to ds:

```js
attachModelToApp() {
   attachModelToDatasource() {
      setupDataAccess() {
       // DefineRelations
       // Here we get the function of each relation method, and wrap them into a function for setRemoting
        Model.relations[relationName] = new relationDefinition();
      }
  }
}

```

A possible way to save relationDefinition without attaching to ds:

```js
RelationDefinition.hasMany = function hasMany(modelFrom, modelTo, params) {
  var relationName = params.as || i8n.camelize(modelTo.pluralModelName, true);
  if (!modelFrom.dataSource) { // not attached yet
     modelFrom.settings.relations[relationName] = {
        // store modelTo, params, etc.
     };
    return;
  }

  // continue as before
  var thisClassName = modelFrom.modelName;
  params = params || {};
  modelTo = lookupModelTo(modelFrom, modelTo, params, true);
```

For the swagger file defined in apic project, since we already set `lazyConnect` as true, the model will always be attached to its ds without connection, which means even ds config is not valid, we can still have the relation apis generated in swagger.

I am closing this issue and the related PR since our current implementation is fine 👌
